docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+
+### Handling of cache refresh
+
+When we create the (singleton) instance of the NetworkServer.LoRaDeviceCache, we start a background
+periodical task to ensure the device twins for all the devices that connected to that LNS are kept
+fresh. In the case where we need to get a device twin, this could trigger a connection ping-pong.
+
+#### Decision
+
+For now we don't make changes here and accept there will be a potential connection switching
+periodically (currently every 2 days if the device was not refreshed in the meantime). This will be
+revisited with either [this issue about the frame counter
+  drift](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1499)
+ (one of the solutions there is to fetch twin and close connection immediately) or [this issue
+ specific to cache refresh](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1505)
+
+#### Alternatives
+
+- Contact the Function to see whether we are the owning gateway. If we are, refresh the twin (and
+  keep the connection open). If we are not:
+  - We adjust the LastSeen property but not actually refresh the entry in the cache.
+    When the next data message for this devices comes in and an entry is in cache for such device, in the
+    event that the Azure Function is marking our LNS as the new "winning" gateway, we have a stale
+    twin in the cache that needs to be updated before validating the request (see [main data flow](#main-data-message-flow)).
+  - Alternatively, we remove device entry from the cache. This idea is discarded because if we were
+    to do this, a "get twin" operation would be triggered as soon as the next data message is coming
+    in from the device in question (for the resolution of devAddr).
+  - We still refresh the twin and close the connection immediately.
+- Remove the background refresh feature all-together: discarded because solutions are possible and
+  not overly complex.
+
+### Handling of Join requests
+
+Join requests in isolation currently do not have the connection stealing issue, as they already rely on the
+Function for the DevNonce check. If the current LNS is not the preferred gateway, it drops the
+message immediately.
+
+Problem
+
+- Join received from LNS1 and LNS2: LNS1 wins
+- First data message received only on LNS2, Function should inform LNS1 that it's the losing one
+  (information about the owning LNS needs to be "shared" between joins and data messages).
+
+#### Decision
+
+We decide to go with the simpler solution of closing the connection immediately after a Join and let
+the data flow re-establish it if needs to be. Disadvantage is that the first data message is more likely
+  to miss the window due to having re-establish the connection.
+  LNS also stores locally that it was the losing LNS and delays on a future re-join or future data messages.
+
+#### Alternative
+
+- Do nothing and accept that a one-off connection stealing on the first Data message can happen.
```

Comment:
This would only work, if you could guarantee that the first data message reaches the gateway the device joined on. Since we can't, we can't rely on the data flow to fix this problem.

Reply:
The only alternative I see here is that during the nonce evaluation, we mark the owning gateway in the function and track it that way. Due to added complexity we discarded that though.